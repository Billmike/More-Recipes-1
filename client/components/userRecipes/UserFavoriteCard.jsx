import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import ReactTooltip from 'react-tooltip';

import RecipeImage from '../common/RecipeImage';

const propTypes = {
  handleOpenDelete: PropTypes.func.isRequired,
  recipe: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    recipeImage: PropTypes.string,
    views: PropTypes.number,
    upvotes: PropTypes.number,
    downvotes: PropTypes.number,
    favorites: PropTypes.number,
  }).isRequired,
  owner: PropTypes.shape({
    fullName: PropTypes.string,
  }).isRequired,
  currentProfileUserId: PropTypes.number.isRequired,
  authenticatedUserId: PropTypes.number,
};

const defaultProps = {
  authenticatedUserId: null,
};

/**
 * User favorite card component
 *
 * @param {object} props - The properties passed to the component
 *
 * @returns {object} React element
 */
const UserFavoriteCard = (props) => {
  const {
    handleOpenDelete, recipe, owner: { fullName }, currentProfileUserId,
    authenticatedUserId
  } = props;
  const { id, title, description, recipeImage, views, upvotes, downvotes,
    favorites } = recipe;
  const message = 'Are you sure you want to remove recipe from your favorites?';
  const actionTitle = 'Remove Recipe';
  const action = 'Remove';

  return (
    <div className="col s12 m6 l3">
      <div id="recipes" className="card large hoverable">
        <a href="recipe.html">
          <span className="card-title">{title}</span>
        </a>
        <div className="card-image">
          <Link to={`/recipes/${id}`}>
            <RecipeImage title={title} recipeImage={recipeImage} />
          </Link>
        </div>
        <div className="card-content">
          <p>{description}</p>
        </div>
        <div className="card-action">
          <p>
            <label htmlFor="owner">Recipe by: </label>
            {fullName}
          </p>
          <ul className="center-align">
            <li id="views"><i className="material-icons tiny">visibility</i>
              {views}
            </li>
            <li id="favorites"><i className="material-icons tiny">favorite</i>
              {favorites}
            </li>
            <li id="upvotes"><i className="material-icons tiny">thumb_up</i>
              {upvotes}
            </li>
            <li id="downvotes"><i className="material-icons tiny">thumb_down</i>
              {downvotes}
            </li>
          </ul>
          {currentProfileUserId === authenticatedUserId &&
            <a
              id="remove-favorite"
              role="button"
              tabIndex="0"
              onClick={() => handleOpenDelete(id, message, actionTitle, action)}
              data-tip="Remove"
            >
              <i className="material-icons icon-red right">remove_circle</i>
            </a>
          }
          <ReactTooltip />
        </div>
      </div>
    </div>
  );
};

UserFavoriteCard.propTypes = propTypes;
UserFavoriteCard.defaultProps = defaultProps;

export default UserFavoriteCard;
