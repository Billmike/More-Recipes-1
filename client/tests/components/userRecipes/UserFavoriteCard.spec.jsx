import React from 'react';
import { BrowserRouter } from 'react-router-dom';

import UserFavoriteCard from '../../../components/userRecipes/UserFavoriteCard';

describe('UserFavoriteCard component', () => {
  it('should render component', () => {
    const props = {
      handleOpenDelete: jest.fn(),
      recipe: {
        title: 'Beans',
        category: 'Main',
        description: 'Tasty beans',
        preparationTime: 85,
        ingredients: 'Beans, onions, palm oil, salt',
        directions: 'Do this. Do that',
      },
      owner: {
        fullName: 'John Mark',
      },
      currentProfileUserId: 1,
      authenticatedUserId: null,
    };
    const wrapper = mount(
      <BrowserRouter><UserFavoriteCard {...props} /></BrowserRouter>
    );

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
  it('should simulate click action', () => {
    const props = {
      handleOpenDelete: jest.fn(),
      recipe: {
        title: 'Beans',
        category: 'Main',
        description: 'Tasty beans',
        preparationTime: 85,
        ingredients: 'Beans, onions, palm oil, salt',
        directions: 'Do this. Do that',
      },
      owner: {
        fullName: 'John Mark',
      },
      currentProfileUserId: 1,
      authenticatedUserId: 1,
    };
    const wrapper = mount(
      <BrowserRouter><UserFavoriteCard {...props} /></BrowserRouter>
    );

    expect(wrapper).toBeDefined();
    wrapper.find('a#remove-favorite').simulate('click');
    expect(props.handleOpenDelete).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
});
