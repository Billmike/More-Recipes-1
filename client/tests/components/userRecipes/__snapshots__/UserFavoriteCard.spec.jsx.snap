// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserFavoriteCard component should render component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "blank",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <UserFavoriteCard
      authenticatedUserId={null}
      currentProfileUserId={1}
      handleOpenDelete={[MockFunction]}
      owner={
        Object {
          "fullName": "John Mark",
        }
      }
      recipe={
        Object {
          "category": "Main",
          "description": "Tasty beans",
          "directions": "Do this. Do that",
          "ingredients": "Beans, onions, palm oil, salt",
          "preparationTime": 85,
          "title": "Beans",
        }
      }
    >
      <div
        className="col s12 m6 l3"
      >
        <div
          className="card large hoverable"
          id="recipes"
        >
          <a
            href="recipe.html"
          >
            <span
              className="card-title"
            >
              Beans
            </span>
          </a>
          <div
            className="card-image"
          >
            <Link
              replace={false}
              to="/recipes/undefined"
            >
              <a
                href="/recipes/undefined"
                onClick={[Function]}
              >
                <RecipeImage
                  recipeImage={null}
                  title="Beans"
                >
                  <img
                    alt="Beans"
                    src={null}
                  />
                </RecipeImage>
              </a>
            </Link>
          </div>
          <div
            className="card-content"
          >
            <p>
              Tasty beans
            </p>
          </div>
          <div
            className="card-action"
          >
            <p>
              <label
                htmlFor="owner"
              >
                Recipe by: 
              </label>
              John Mark
            </p>
            <ul
              className="center-align"
            >
              <li
                id="views"
              >
                <i
                  className="material-icons tiny"
                >
                  visibility
                </i>
              </li>
              <li
                id="favorites"
              >
                <i
                  className="material-icons tiny"
                >
                  favorite
                </i>
              </li>
              <li
                id="upvotes"
              >
                <i
                  className="material-icons tiny"
                >
                  thumb_up
                </i>
              </li>
              <li
                id="downvotes"
              >
                <i
                  className="material-icons tiny"
                >
                  thumb_down
                </i>
              </li>
            </ul>
            <ReactTooltip
              insecure={true}
              resizeHide={true}
              wrapper="div"
            >
              <div
                className="__react_component_tooltip place-top type-dark "
                data-id="tooltip"
              />
            </ReactTooltip>
          </div>
        </div>
      </div>
    </UserFavoriteCard>
  </Router>
</BrowserRouter>
`;

exports[`UserFavoriteCard component should simulate click action 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "blank",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <UserFavoriteCard
      authenticatedUserId={1}
      currentProfileUserId={1}
      handleOpenDelete={
        [MockFunction] {
          "calls": Array [
            Array [
              undefined,
              "Are you sure you want to remove recipe from your favorites?",
              "Remove Recipe",
              "Remove",
            ],
          ],
        }
      }
      owner={
        Object {
          "fullName": "John Mark",
        }
      }
      recipe={
        Object {
          "category": "Main",
          "description": "Tasty beans",
          "directions": "Do this. Do that",
          "ingredients": "Beans, onions, palm oil, salt",
          "preparationTime": 85,
          "title": "Beans",
        }
      }
    >
      <div
        className="col s12 m6 l3"
      >
        <div
          className="card large hoverable"
          id="recipes"
        >
          <a
            href="recipe.html"
          >
            <span
              className="card-title"
            >
              Beans
            </span>
          </a>
          <div
            className="card-image"
          >
            <Link
              replace={false}
              to="/recipes/undefined"
            >
              <a
                href="/recipes/undefined"
                onClick={[Function]}
              >
                <RecipeImage
                  recipeImage={null}
                  title="Beans"
                >
                  <img
                    alt="Beans"
                    src={null}
                  />
                </RecipeImage>
              </a>
            </Link>
          </div>
          <div
            className="card-content"
          >
            <p>
              Tasty beans
            </p>
          </div>
          <div
            className="card-action"
          >
            <p>
              <label
                htmlFor="owner"
              >
                Recipe by: 
              </label>
              John Mark
            </p>
            <ul
              className="center-align"
            >
              <li
                id="views"
              >
                <i
                  className="material-icons tiny"
                >
                  visibility
                </i>
              </li>
              <li
                id="favorites"
              >
                <i
                  className="material-icons tiny"
                >
                  favorite
                </i>
              </li>
              <li
                id="upvotes"
              >
                <i
                  className="material-icons tiny"
                >
                  thumb_up
                </i>
              </li>
              <li
                id="downvotes"
              >
                <i
                  className="material-icons tiny"
                >
                  thumb_down
                </i>
              </li>
            </ul>
            <a
              data-tip="Remove"
              onClick={[Function]}
              role="button"
              tabIndex="0"
            >
              <i
                className="material-icons icon-red right"
                id="remove-favorite"
              >
                remove_circle
              </i>
            </a>
            <ReactTooltip
              insecure={true}
              resizeHide={true}
              wrapper="div"
            >
              <div
                className="__react_component_tooltip place-top type-dark "
                data-id="tooltip"
              />
            </ReactTooltip>
          </div>
        </div>
      </div>
    </UserFavoriteCard>
  </Router>
</BrowserRouter>
`;
